contributors/design-proposals/termination-reason.md

```diff
@@ -0,0 +1,317 @@
+
+DeleteOptions is modified to carry a reason.
+
+```golang
+type DeleteOptions struct {
+	// Other fields omitted for brevity.
+	
+	// Reason indicates the reason for the deletion of an API Object that 
+	// undergo graceful termination upon deletion.
+	Reason string `json:"reason,omitempty"`
+}
+```
+
+The ObjectMeta struct is modified to carry the deletion reason via 
+an Object's metadata.
+
+```golang
+type ObjectMeta {
+	// Other fields omitted for brevity.
+	
+    // DeletionReason indicates the reason for the deletion of an API Object. 
+    // Its purpose is to provide an extra generic signal to watchers of API 
+    // Objects during the graceful termination process. 
+	DeletionReason string `json:"reason,omitempty"`
+}
+```
+
+DeleteExecAction and DeleteHttpAction are introduced to provide users with 
+the ability to configure the environment variable or HTTP Header that will be 
+used to convey a termination reason signaled by a DeletionReason.
+
+```golang
+// DeleteExecAction describes a "run in container" action that will be invoked
+// invoked inside of a container prior to sending the TERM signal to the 
+// container's entry point.
+type DeleteExecAction struct {
+	// Command is the command line to execute inside the container, the working directory for the
+	// command  is root ('/') in the container's filesystem.  The command is simply exec'd, it is
+	// not run inside a shell, so traditional shell instructions ('|', etc) won't work.  To use
+	// a shell, you need to explicitly call out to that shell.
+	// +optional
+	Command []string
+	
+	// ReasonEnv is the environment variable that wil be populated with the 
+	// reason, if provided, for the Pod's termination. This variable defaults
+	// to "KUBE_DELETE_REASON"
+	ReasonEnv string
+}
+
+// DeleteHTTPGetAction describes an action to take upon deletion based on HTTP 
+// Get requests.
+type DeleteHTTPGetAction struct {
+	// Optional: Path to access on the HTTP server.
+	// +optional
+	Path string
+	// Required: Name or number of the port to access on the container.
+	// +optional
+	Port intstr.IntOrString
+	// Optional: Host name to connect to, defaults to the pod IP. You
+	// probably want to set "Host" in httpHeaders instead.
+	// +optional
+	Host string
+	// Optional: Scheme to use for connecting to the host, defaults to HTTP.
+	// +optional
+	Scheme URIScheme
+	// Optional: Custom headers to set in the request. HTTP allows repeated headers.
+	// +optional
+	HTTPHeaders []HTTPHeader
+	// ReasonHeader is the header that will be set to the reason for a 
+	// deletion. This header defaults to "KUBE-DELETE-REASON"
+	ReasonHeader string
+}
+```
+
+PreStopHandler is introduced to specialize the current Handler implementation 
+to provide a termination reason during the execution of Lifecycle's PreStop 
+handler.
+
+```golang
+// PreStopHandler invokes either a DeleteExecAction or a DeleteHTTPGetAction 
+// prior to the graceful termination of a Pod.
+type PreStopHandler struct {
```

Comment:
I'm confused why this isn't PreDeleteHandler, or StopExecAction.  Shouldn't be both.

Reply:
left this as PreStopHandler and changed the variables internally yo StopExecAction, StopHTTPGetAction